tests/topotests/bgp_bmp/test_bgp_bmp.py

```diff
@@ -0,0 +1,324 @@
+            TopoRouter.RD_BGP,
+            os.path.join(CWD, "{}/bgpd.conf".format(rname)),
+            "-M bmp",
+        )
+
+    tgen.start_router()
+
+    logger.info("starting BMP servers")
+    for pname, server in tgen.get_pmacct_servers().items():
+        server.start(os.path.join(CWD, pname))
+
+
+def teardown_module(_mod):
+    tgen = get_topogen()
+    tgen.stop_topology()
+
+
+def test_bgp_convergence():
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    result = verify_bgp_convergence_from_running_config(tgen, dut="r1")
+    assert result is True, "BGP is not converging"
+
+
+def get_bmp_messages():
+    """
+    Read the BMP logging messages.
+    """
+    messages = []
+    tgen = get_topogen()
+    text_output = tgen.gears["bmp1"].run("cat /var/log/bmp-192_0_2_1.log")
+
+    for m in text_output.splitlines():
+        # some output in the bash can break the message decoding
+        try:
+            messages.append(json.loads(m))
+        except Exception as e:
+            logger.warning(str(e) + " message: {}".format(str(m)))
+            continue
+
+    if not messages:
+        logger.error("Bad BMP log format, check your BMP server")
+
+    return messages
+
+
+def check_for_prefixes(expected_prefixes, msg_type, policy):
+    """
+    Check for the presence of the given prefixes in the BMP server logs with
+    the given message type and the set policy.
+    """
+    global SEQ
+    # we care only about the new messages
+    messages = [
+        m for m in sorted(get_bmp_messages(), key=lambda d: d["seq"]) if m["seq"] > SEQ
+    ]
+
+    # get the list of pairs (prefix, policy, seq) for the given message type
+    p_flag = "is_post" if policy == "is_post" else "is_in"
+    prefix_policy = [
+        (m["ip_prefix"], m[p_flag], m["seq"])
+        for m in messages
+        if "log_type" in m.keys() and m["log_type"] == msg_type
+    ]
+
+    # check for prefixes
+    prefixes = [p[0] for p in prefix_policy]
+    for ep in expected_prefixes:
+        if ep not in prefixes:
+            msg = "The prefix {} is not present in the {} log messages."
+            logger.debug(msg.format(ep, msg_type))
+            return False
+
+    # check for policy flag, (policy, seq)
+    policies = [(p[1], p[2]) for p in prefix_policy]
+    for pf in policies:
+        if not int(pf[0]) == 1:
+            msg = "The BMP message {} has not the right message policy type {}"
+            logger.debug(msg.format(pf[1], policy))
+            return False
+
+    SEQ = messages[-1]["seq"]
+
+
+def ipv4_v6_unicast_prefixes(policy):
+    """
+    Setup the BMP  monitor policy, Add and withdraw ipv4/v6 prefixes.
+    Check if the previous actions are logged in the BMP server with the right
+    message type and the right policy.
+    """
+    tgen = get_topogen()
+    # configure the bmp policy on the r1 router
+    del_policy = "post-policy" if policy == "pre-policy" else "pre-policy"
```

Comment:
using an enum for different monitoring types would be better and more extensible for future monitoring types (loc-rib and rib-out)